src/Microsoft.VisualStudio.ProjectSystem.Managed/LinqExtensions.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.VisualStudio
+{
+    internal static class LinqExtensions
+    {
+        [Pure]
+        public static T FirstOrDefault<T, TArg>(this IEnumerable<T> source, Func<T, TArg, bool> predicate, TArg arg)
+        {
+            foreach (T obj in source)
+            {
+                if (predicate(obj, arg))
+                    return obj;
+            }
+
+            return default;
+        }
+
+        [Pure]
+        public static T SingleOrDefault<T, TArg>(this IEnumerable<T> source, Func<T, TArg, bool> predicate, TArg arg)
+        {
+            T match = default;
+            long count = 0;
+
+            foreach (T item in source)
+            {
+                if (predicate(item, arg))
+                {
+                    match = item;
+                    checked { ++count; }
+                }
```

Comment:
Rationale here was to avoid check in the happy case, as if you're going to throw you have bigger problems to worry about. But looking at it again, I think there's a nicer implementation.